chia/util/keychain.py

```diff
@@ -337,43 +360,58 @@ def _get_free_private_key_index(self) -> int:
             except KeychainUserNotFound:
                 return index
 
+    # pylint requires these NotImplementedErrors for some reason
     @overload
-    def add_key(self, mnemonic_or_pk: str) -> PrivateKey: ...
+    def add_key(self, mnemonic_or_pk: str) -> Tuple[PrivateKey, KeyTypes]:
+        raise NotImplementedError()  # pragma: no cover
 
     @overload
-    def add_key(self, mnemonic_or_pk: str, label: Optional[str]) -> PrivateKey: ...
+    def add_key(self, mnemonic_or_pk: str, label: Optional[str]) -> Tuple[PrivateKey, KeyTypes]:
+        raise NotImplementedError()  # pragma: no cover
 
     @overload
-    def add_key(self, mnemonic_or_pk: str, label: Optional[str], private: Literal[True]) -> PrivateKey: ...
+    def add_key(self, mnemonic_or_pk: str, label: Optional[str], private: Literal[True]) -> Tuple[PrivateKey, KeyTypes]:
+        raise NotImplementedError()  # pragma: no cover
 
     @overload
-    def add_key(self, mnemonic_or_pk: str, label: Optional[str], private: Literal[False]) -> G1Element: ...
+    def add_key(
+        self, mnemonic_or_pk: str, label: Optional[str], private: Literal[False]
+    ) -> Tuple[ObservationRoot, KeyTypes]:
+        raise NotImplementedError()  # pragma: no cover
 
     @overload
-    def add_key(self, mnemonic_or_pk: str, label: Optional[str], private: bool) -> Union[PrivateKey, G1Element]: ...
+    def add_key(
+        self, mnemonic_or_pk: str, label: Optional[str], private: bool
+    ) -> Tuple[Union[PrivateKey, ObservationRoot], KeyTypes]:
+        raise NotImplementedError()  # pragma: no cover
 
     def add_key(
         self, mnemonic_or_pk: str, label: Optional[str] = None, private: bool = True
-    ) -> Union[PrivateKey, G1Element]:
+    ) -> Tuple[Union[PrivateKey, ObservationRoot], KeyTypes]:
         """
         Adds a key to the keychain. The keychain itself will store the public key, and the entropy bytes (if given),
         but not the passphrase.
         """
-        key: Union[PrivateKey, G1Element]
+        key: Union[PrivateKey, ObservationRoot]
+        key_type: KeyTypes
         if private:
             seed = mnemonic_to_seed(mnemonic_or_pk)
             entropy = bytes_from_mnemonic(mnemonic_or_pk)
             index = self._get_free_private_key_index()
             key = AugSchemeMPL.key_gen(seed)
+            key_type = KeyTypes.G1_ELEMENT
             assert isinstance(key, PrivateKey)
             pk = key.get_g1()
             key_data = bytes(pk).hex() + entropy.hex()
             fingerprint = pk.get_fingerprint()
         else:
             index = self._get_free_private_key_index()
             pk_bytes = hexstr_to_bytes(mnemonic_or_pk)
-            key = G1Element.from_bytes(pk_bytes)
-            assert isinstance(key, G1Element)
+            if len(pk_bytes) == 48:
```

Comment:
This kind of length check doesn't thrill me , I'm not sure it's even needed - I expect you could just call G1Element.from_byte in a try?

Reply:
I was under the impression that using try/catch for control flow was bad practice.  And under the hood G1Element is going to do this anyways.  I'm not thrilled with the length check either but it seems to be the best way we have available to us with existing keyring format not having anticipated further types.